### PR TITLE
Merge paginated plan optimizer into the regular optimizer.

### PR DESCRIPTION
**Pull request by MaxKsyunz**
Address [this](https://github.com/opensearch-project/sql/pull/1497#discussion_r1160959892) comment in https://github.com/opensearch-project/sql/pull/1497

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: core/src/test/java/org/opensearch/sql/planner/optimizer/LogicalPlanOptimizerTest.java [hunk 1]
```diff
@@ -313,48 +326,76 @@ public PhysicalPlan implement(LogicalPlan plan) {
 
   @Test
   void paged_table_scan_builder_support_project_push_down_can_apply_its_rule() {
-    when(tableScanBuilder.pushDownProject(any())).thenReturn(true);
-    when(table.createPagedScanBuilder(anyInt())).thenReturn(tableScanBuilder);
 
-    var relation = new LogicalRelation("schema", table);
-    relation.setPageSize(anyInt());
+    var relation = Mockito.spy(new LogicalRelation("schema", table));
```
Comment (Yury-Fridlyand) [inline core/src/test/java/org/opensearch/sql/planner/optimizer/LogicalPlanOptimizerTest.java:330]: Maybe `spy` is not needed

File: core/src/test/java/org/opensearch/sql/planner/optimizer/LogicalPlanOptimizerTest.java [hunk 2]
```diff
@@ -62,11 +67,19 @@ class LogicalPlanOptimizerTest {
...
   @BeforeEach
   void setUp() {
     lenient().when(table.createScanBuilder()).thenReturn(tableScanBuilder);
+    lenient().when(table.createPagedScanBuilder(anyInt())).thenReturn(pagedTableScanBuilder);
   }
 
+
+
+
+
```
Comment (Yury-Fridlyand) [inline core/src/test/java/org/opensearch/sql/planner/optimizer/LogicalPlanOptimizerTest.java:82]:
```suggestion
```

File: core/src/main/java/org/opensearch/sql/planner/optimizer/rule/CreatePagingTableScanBuilder.java [hunk 3]
```diff
@@ -5,45 +5,59 @@
...
+
   @Override
-  public LogicalPlan apply(LogicalRelation plan, Captures captures) {
-    TableScanBuilder scanBuilder = captures.get(capture)
-        .createPagedScanBuilder(plan.getPageSize());
-    // TODO: Remove this after Prometheus refactored to new table scan builder too
-    return (scanBuilder == null) ? plan : scanBuilder;
+  public LogicalPlan apply(LogicalPaginate plan, Captures captures) {
+    var logicalRelation = (LogicalRelation) relationParent.getChild().get(0);
+    var scan = logicalRelation.getTable().createPagedScanBuilder(logicalRelation.getPageSize());
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/planner/optimizer/rule/CreatePagingTableScanBuilder.java:58]: If we know (and set) page size here, we can omit `PushPageSize` then.

File: core/src/main/java/org/opensearch/sql/planner/optimizer/rule/CreatePagingTableScanBuilder.java [hunk 4]
```diff
@@ -5,45 +5,59 @@
...
+  private final Pattern<LogicalPaginate> pattern;
 
   /**
    * Constructor.
    */
   public CreatePagingTableScanBuilder() {
-    this.capture = Capture.newCapture();
-    this.pattern = Pattern.typeOf(LogicalRelation.class)
-        .with(table().capturedAs(capture));
+    this.pattern = Pattern.typeOf(LogicalPaginate.class).matching(lp -> {
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/planner/optimizer/rule/CreatePagingTableScanBuilder.java:36]: Maybe: leave a comment describing what happens here

File: core/src/main/java/org/opensearch/sql/executor/QueryService.java [hunk 5]
```diff
@@ -34,11 +32,9 @@ public class QueryService {
    * There are two planners, one - to handle pagination requests (cursor/scroll) only and
    * another one for everything else.
    * @see OpenSearchPluginModule#queryPlanFactory (:plugin module)
-   * @see LogicalPlanOptimizer#paginationCreate
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/executor/QueryService.java:37]: Please, update comment above

File: core/src/main/java/org/opensearch/sql/planner/optimizer/rule/CreatePagingTableScanBuilder.java (hunk 3 above)
Comment (MaxKsyunz) [reply core/src/main/java/org/opensearch/sql/planner/optimizer/rule/CreatePagingTableScanBuilder.java:58]: True. Would this make your push size PR unnecessary?

Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/planner/optimizer/rule/CreatePagingTableScanBuilder.java:58]: Yes, and good if so.

**Review by GabeFernandez310 — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/planner/optimizer/rule/CreatePagingTableScanBuilder.java [hunk 6]
```diff
@@ -5,45 +5,66 @@
...
+    do {
+      var plan = plans.pop();
+      if (plan.getChild().stream().anyMatch(LogicalRelation.class::isInstance)) {
+        if (plan.getChild().size() > 1) {
+          throw new UnsupportedOperationException();
+        }
+        relationParent = plan;
+        return true;
+      }
+      plan.getChild().forEach(plans::push);
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/planner/optimizer/rule/CreatePagingTableScanBuilder.java:56]: There's an addAll() but no pushAll() method... hmm
Maybe use `addAll()` and `removeFirst()` instead?

**Review by acarbonetto — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/planner/optimizer/rule/CreatePagingTableScanBuilder.java (hunk 6 above)
Comment (MaxKsyunz) [reply core/src/main/java/org/opensearch/sql/planner/optimizer/rule/CreatePagingTableScanBuilder.java:56]: I'll switch for the sake of removing `forEach(plans:push)`.

File: core/src/main/java/org/opensearch/sql/planner/optimizer/rule/CreatePagingTableScanBuilder.java [hunk 7]
```diff
@@ -5,45 +5,66 @@
...
+  }
+
+  /**
+   * Finds an instance of LogicalRelation and saves a reference in relationParent variable.
+   * @param logicalPaginate An instance of LogicalPaginate
+   * @return true if {@link LogicalRelation} node was found among the descendents of
+   *     {@link this.logicalPaginate}, false otherwise.
+   */
+  private boolean findLogicalRelation(LogicalPaginate logicalPaginate) {
+    Deque<LogicalPlan> plans = new ArrayDeque<>();
```
[the remark this answers is not in the document]
Comment (MaxKsyunz) [reply core/src/main/java/org/opensearch/sql/planner/optimizer/rule/CreatePagingTableScanBuilder.java:45]: Sadly, I cannot -- that constructor expects a collection.

Compared to adding a conversion to collection,  just calling `push` is simpler.

**Pull request merged**